vaccine/ask_a_question.py

```diff
@@ -0,0 +1,237 @@
+        )
+
+        async def check(text: Optional[str]):
+            await nonempty_validator(question)(text)
+            # overwite state_question with this new question
+            self.save_answer("state_question", text)
+
+        return FreeText(self, question=question, check=check, next="state_call_model")
+
+    async def state_display_response_choices(self):
+        class RedirectChoiceState(ChoiceState):
+            async def process_message(self, message: Message):
+                choice = self._get_choice(message.content)
+                if choice is None:
+                    state = await self.app.go_to_state("state_question")
+                    return await state.process_message(message)
+                else:
+                    return await super().process_message(message)
+
+        responses = json.loads(self.user.answers["model_response"])["top_responses"]
+        return RedirectChoiceState(
+            self,
+            question=self._("🔎 *Top Search Results*\n"),
+            choices=[Choice(title, title) for title, _ in responses],
+            error="",  # Errors now redirect to `state_question`
+            next="state_display_selected_choice",
+            footer=self._(
+                "\n[💡Tip: If you don't see what you're looking for, try typing your "
+                "question again using different words or reply *FAQ* to browse topics]"
+            ),
+        )
+
+    async def state_display_selected_choice(self):
+        responses = json.loads(self.user.answers["model_response"])["top_responses"]
+        choice = self.user.answers["state_display_response_choices"]
+        for title, content in responses:
+            if choice == title:
+                break
+        question = self._(
+            "*{title}*\n"
+            "\n"
+            "{content}\n"
+            "\n"
+            "------------\n"
+            "_Help us get better and finding answers for you._\n"
+            "Did the information above ☝🏽 answer your question?\n"
+            "Reply:"
+        ).format(title=title, content=content)
+        return ChoiceState(
+            self,
+            question=question,
+            choices=[Choice("yes", self._("*YES*")), Choice("no", self._("*NO*"))],
+            error=question,
+            next="state_submit_user_feedback",
+        )
+
+    async def state_submit_user_feedback(self):
+        model = get_model()
+        model_response = json.loads(self.user.answers["model_response"])
+        data = {
+            "inbound_id": model_response["inbound_id"],
+            "feedback_secret_key": model_response["feedback_secret_key"],
+            "feedback": self.user.answers["state_display_selected_choice"],
+        }
+        async with model as session:
+            for i in range(3):
+                try:
+                    response = await session.post(
+                        url=urljoin(config.MODEL_API_URL, "/inbound/feedback"),
```

Comment:
This is cool, do they use it for training?

Reply:
Yeah, but it's not for automated training. This just stores it in the database so that we can pull it later, and use it to improve the model.